pkg/controller/controllerutil/controllerutil.go

```diff
@@ -46,3 +50,56 @@ func SetControllerReference(owner, object v1.Object, scheme *runtime.Scheme) err
 	object.SetOwnerReferences(append(object.GetOwnerReferences(), ref))
 	return nil
 }
+
+// OperationType is the action result of a CreateOrUpdate call
+type OperationType string
+
+const ( // They should complete the sentence "Deployment default/foo has been ..."
+	// OperationNoop means that the resource has not been changed
+	OperationNoop = "unchanged"
+	// OperationCreated means that a new resource has been created
+	OperationCreated = "created"
+	// OperationUpdated means that an existing resource has been updated
+	OperationUpdated = "updated"
+)
+
+// CreateOrUpdate creates or updates a kuberenes resource. It takes in a key and
+// a placeholder for the existing object and returns the modified object
+func CreateOrUpdate(ctx context.Context, c client.Client, key client.ObjectKey, existing runtime.Object, t TransformFn) (runtime.Object, OperationType, error) {
+	err := c.Get(ctx, key, existing)
+	var obj runtime.Object
+
+	if errors.IsNotFound(err) {
+		obj, err = t(existing)
+		if err != nil {
+			return nil, OperationNoop, err
+		}
+
+		err = c.Create(ctx, obj)
+
+		if err != nil {
+			return nil, OperationNoop, err
+		}
+		return obj, OperationCreated, err
+	} else if err != nil {
+		return nil, OperationNoop, err
+	} else {
+		obj, err = t(existing.DeepCopyObject())
+		if err != nil {
+			return nil, OperationNoop, err
+		}
+
+		if !reflect.DeepEqual(existing, obj) {
+			err = c.Update(ctx, obj)
+			if err != nil {
+				return nil, OperationNoop, err
+			}
+			return obj, OperationUpdated, err
+		}
+		return obj, OperationNoop, nil
+	}
+}
+
+// TransformFn is a function which take in a kubernetes object and returns the
+// desired state of that object
```

Comment:
note whether or not it's allowed/expected to mutate the object.  It can be helpful when refactoring later.